src/PluginsManager/Microsoft.Performance.Toolkit.Plugins.Runtime/Manager/PluginsSystem.cs

```diff
@@ -0,0 +1,114 @@
+        private PluginsSystem(
+            IPluginsInstaller installer,
+            IPluginsDiscoverer discoverer,
+            IRepository<PluginSource> pluginSourceRepository,
+            IRepositoryRO<IPluginDiscovererProvider> pluginDiscovererProviderRepository,
+            IRepositoryRO<IPluginFetcher> pluginFetcherRepository)
+        {
+            this.Installer = installer;
+            this.Discoverer = discoverer;
+            this.PluginSourcesRepository = pluginSourceRepository;
+            this.DiscovererProvidersRepository = pluginDiscovererProviderRepository;
+            this.FetchersRepository = pluginFetcherRepository;
+        }
+
+        /// <summary>
+        ///     Creates a file based plugins system.
+        /// </summary>
+        /// <param name="root">
+        ///     The root directory of the registry and installed plugins.
+        /// </param>
+        /// <param name="pluginSourcesRepo">
+        ///     The repository of plugin sources.
+        /// </param>
+        /// <param name="fetchersRepo">
+        ///     The repository of fetchers.
+        /// </param>
+        /// <param name="discovererProvidersRepo">
+        ///     The repository of discoverer providers.
+        /// </param>
+        /// <param name="logger">
+        ///     The logger to use.
+        /// </param>
+        /// <returns>
+        ///     The created plugins system.
+        /// </returns>
+        public static PluginsSystem CreateFileBasedPluginsSystem(
+            string root,
+            ILogger logger)
+        {
+            IRepository<PluginSource> pluginSourcesRepo = new PluginSourceRepository();
+            IRepositoryRO<IPluginFetcher> fetchersRepo = new PluginsManagerResourceRepository<IPluginFetcher>();
+            IRepositoryRO<IPluginDiscovererProvider> discovererProvidersRepo = new PluginsManagerResourceRepository<IPluginDiscovererProvider>();
+
+            IPluginsDiscoverer discoverer = new PluginsDiscoverer(
+              pluginSourcesRepo,
+              fetchersRepo,
+              discovererProvidersRepo,
+              logger);
+
+
+            IPluginRegistry registry = new FileBackedPluginRegistry(root);
+            IPluginsInstaller installer = new FileBackedPluginsInstaller(root, registry);
+          
+            return new PluginsSystem(
+                installer,
+                discoverer,
+                pluginSourcesRepo,
+                discovererProvidersRepo,
+                fetchersRepo);
+        }
+
+        /// <summary>
+        ///     Gets the installer.
+        /// </summary>
+        public IPluginsInstaller Installer { get; }
+
+        /// <summary>
+        ///     Gets the discoverer.
+        /// </summary>
+        public IPluginsDiscoverer Discoverer { get; }
+
+        /// <summary>
+        ///     Gets the repository of plugin fetchers.
+        /// </summary>
+        public IRepositoryRO<IPluginFetcher> FetchersRepository { get; }
```

Comment:
Do we need this (and the one below)? These seem like an implementation detail

Reply:
Actually I just realized this won't work. Because we are creating these repositories in the static method hot path and only exposing a readonly interface, there's no way for the client to pass any dicoverers and fectchers down to the plugins system.   

I think we should either use another interface (supports adding new resource and interacts with resource loader) so that resources can be added later by the client, or have the client initialize these repositories. What do you think?

Reply:
Yeah, I think we may want to extract out and expose some type an `IPluginSystemResourceLoader` interface for consumers to "load" instances of discoverers/fetchers. The loader would be passed down to anyone who needs the resources as an `IRepositoryRO<T>`. For example, we may want to have:

```cs
public interface IPluginSystemResourcesLoader
{
    void Add(IPluginDiscoverer discoverer);

    void Add(IPluginFetcher fetcher);
}

public class PluginSystemResourcesLoader
    : IPluginSystemResourcesLoader,
      IRepositoryRO<IPluginDiscoverer>,
      IRepositoryRO<IPluginFetcher>
{
    ...
}
```

Reply:
We already have a mechanism  [IPluginsManagerResourceLoader](https://github.com/microsoft/microsoft-performance-toolkit-sdk/blob/feature/pluginManager/src/PluginsManager/Microsoft.Performance.Toolkit.Plugins.Runtime/Extensibility/IPluginsManagerResourceLoader.cs) for loading resources using a subscribe/unsubscribe model similar to what the SDK runtime does today:

The [PluginsManagerResourceRepository](https://github.com/microsoft/microsoft-performance-toolkit-sdk/blob/feature/pluginManager/src/PluginsManager/Microsoft.Performance.Toolkit.Plugins.Runtime/Extensibility/PluginsManagerResourceRepository.cs)  is both an `IRepositoryRO` and an `IPluginsManagerResourcesConsumer` which can be registered to the loader to receive new resources. We can create a new interface that inherits both of them and use it as the type we expose in the `PluginsSystem` class.

Reply:
@mslukebo Pushed an update addressing this.